x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/utils/ExceptionRootCauseFinder.java

```diff
@@ -63,7 +64,7 @@ public static Throwable getFirstIrrecoverableExceptionFromBulkResponses(Collecti
             }
 
             if (unwrappedThrowable instanceof ElasticsearchException elasticsearchException) {
-                if (isExceptionIrrecoverable(elasticsearchException)) {
+                if (isExceptionIrrecoverable(elasticsearchException) && isNotIndexNotFoundException(elasticsearchException)) {
```

Comment:
If we no longer consider `IndexNotFoundException` irrecoverable, could we modify the `isExceptionIrrecoverable` method to reflect that instead of adding the `isNotIndexNotFoundException` method?

Or the distinction is still needed for cases other than bulk responses?

Reply:
That is a good question - at least for now I was thinking of just handling it for the destination index? The source index can still fail with IndexNotFoundException (if PIT == null): https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java#L602

I haven't seen this same type of intermittent failure scenario though, where the source index gets relocated to a new shard and the request fails on the old shard, so I'm not sure if we just haven't seen it yet or if search behaves differently where the request will automatically reroute.

Reply:
> I'm not sure if we just haven't seen it yet or if search behaves differently where the request will automatically reroute.

I don't know either unfortunately :(
But since we are going in a direction to make more errors recoverable (and more transforms unattended) it could make sense to make `IndexNotFoundException` recoverable.
But maybe let's not do it in this PR if you don't feel comfortable.
We can leave this PR for dealing with destination index only.
Please make it clear in the PR title/description.